core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushPartialTopNIntoTableScan.java

```diff
@@ -0,0 +1,277 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.Session;
+import io.trino.matching.Capture;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.metadata.Metadata;
+import io.trino.metadata.TableHandle;
+import io.trino.spi.connector.ColumnHandle;
+import io.trino.spi.connector.PartialSortApplicationResult;
+import io.trino.spi.connector.SortOrder;
+import io.trino.sql.PlannerContext;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.iterative.GroupReference;
+import io.trino.sql.planner.iterative.Lookup;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.optimizations.PlanNodeSearcher;
+import io.trino.sql.planner.plan.ExchangeNode;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.planner.plan.LastNNode;
+import io.trino.sql.planner.plan.LimitNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.PlanVisitor;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.planner.plan.TableScanNode;
+import io.trino.sql.planner.plan.TopNNode;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import static io.trino.SystemSessionProperties.isPushPartialTopNIntoTableScan;
+import static io.trino.sql.planner.plan.Patterns.TopN.step;
+import static io.trino.sql.planner.plan.Patterns.exchange;
+import static io.trino.sql.planner.plan.Patterns.source;
+import static java.lang.Math.toIntExact;
+
+/**
+ * This rule rewrites the partial TopNNode to LimitNode or LastNNode.
+ *
+ * <p>
+ * Transforms:
+ * <pre> {@code
+ *    Top[partial]
+ *      Project
+ *        Filter
+ *          TableScan
+ *        }
+ * </pre>
+ * into:
+ * <pre> {@code
+ *    Limit / LastN
+ *      Project
+ *        Filter
+ *         TableScan
+ *    }
+ *  </pre>
+ *
+ */
+public class PushPartialTopNIntoTableScan
```

Comment:
cc @martint

Reply:
Hi, @martint Just a friendly ping, Could you take a look on this  ?